opt: fix trigger+computed column internal error

mgartner opened the pull request:
This commit fixes a bug that was caused by incorrectly modifying the set
and list of target columns when re-projecting computed columns after
building a `BEFORE` trigger.

Fixes #154672

Release note (bug fix): A bug has been fixed that caused internal errors
for `INSERT .. ON CONFLICT .. DO UPDATE` statements when the target
table had both a computed column and a `BEFORE` trigger. This bug has
been present since triggers were introduced in v24.3.0.

DrewKimball reviewed (commented):
It seems that `targetCols` is only used to keep track of the columns directly being inserted into or being set by an update. I think the right way to fix it is actually to have `addSynthesizedComputedCols` skip appending to `targetCols` when called for a trigger. What do you think?

I think the `mutationBuilder` could use some refactoring. That field is really only meant to be used "locally" while building the input for an insert or an update (and is unset after that's done), but that's not really clear from the comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154789#-:-OafwaLs4TEXY35WBgHT:b-90ge18)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner commented:
>It seems that targetCols is only used to keep track of the columns directly being inserted into or being set by an update. I think the right way to fix it is actually to have addSynthesizedComputedCols skip appending to targetCols when called for a trigger. What do you think?

I like it. I tried something similar on Friday, but I forgot why I abandoned it. Let me try again.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> @DrewKimball reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154789#-:-Oavx9oKCFSVLCdSWEu2:b-g2a5fb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/opt/optbuilder/mutation_builder.go:
```diff
@@ -795,7 +795,22 @@ func (mb *mutationBuilder) addSynthesizedDefaultCols(
+func (mb *mutationBuilder) addSynthesizedComputedCols(colIDs opt.OptionalColList, restrict bool) {
+	mb.targetColSet = mb.projectSynthesizedComputedCols(colIDs, restrict)
+	for col, ok := mb.targetColSet.Next(0); ok; col, ok = mb.targetColSet.Next(col + 1) {
+		mb.targetColList = append(mb.targetColList, col)
```
I think this assumes that table columns have ascending IDs... which is true right now, but maybe not something we want to rely on.

Here's one place where we assume that the order of columns in `targetColList` is the same as the order of columns in `mb.outScope` after `projectSynthesizedComputedCols` returns: https://github.com/cockroachdb/cockroach/blob/b845bd8573dd8787becad6df17424df2772c3940/pkg/sql/opt/optbuilder/insert.go#L699-L701

The need for `targetColList` is dubious... Maybe we can get rid of it in the near future...

In the meantime I'm going to attempt a safer fix.

mgartner committed:
opt: fix trigger+computed column internal error

This commit fixes a bug that was caused by incorrectly modifying the set
and list of target columns when re-projecting computed columns after
building a `BEFORE` trigger.

Fixes cockroachdb#154672

Release note (bug fix): A bug has been fixed that caused internal errors
for `INSERT .. ON CONFLICT .. DO UPDATE` statements when the target
table had both a computed column and a `BEFORE` trigger. This bug has
been present since triggers were introduced in v24.3.0.

mgartner commented:
Ok, third time's the charm. PTAL!

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @DrewKimball reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154789#-:-Ob1Daq-2L2R553egriU:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
